src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+
+    while (true)
+    {
+        // check that the request can ultimately be satisfied.
+        VerifyOrExit(requiredSpace <= eventBuffer->GetQueueSize(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        if (requiredSpace > eventBuffer->AvailableDataLength())
+        {
+            ctx.mpEventBuffer        = eventBuffer;
+            ctx.mSpaceNeededForEvent = 0;
+
+            eventBuffer->mProcessEvictedElement = EvictEvent;
+            eventBuffer->mAppData               = &ctx;
+            err                                 = eventBuffer->EvictHead();
+
+            // one of two things happened: either the element was evicted immediately if the head's priority is same as current
+            // buffer(final one), or we figured out how much space we need to evict it into the next buffer, the check happens in
+            // EvictEvent function
+
+            if (err != CHIP_NO_ERROR)
+            {
+                VerifyOrExit(ctx.mSpaceNeededForEvent != 0, /* no-op, return err */);
+                if (ctx.mSpaceNeededForEvent <= eventBuffer->mpNext->AvailableDataLength())
+                {
+                    // we can copy the event outright.  copy event and
+                    // subsequently evict head s.t. evicting the head
+                    // element always succeeds.
+                    // Since we're calling CopyElement and we've checked
+                    // that there is space in the next buffer, we don't expect
+                    // this to fail.
+                    err = CopyToNextBuffer(eventBuffer);
+                    SuccessOrExit(err);
+                    // success; evict head unconditionally
+                    eventBuffer->mProcessEvictedElement = nullptr;
+                    err                                 = eventBuffer->EvictHead();
+                    // if unconditional eviction failed, this
+                    // means that we have no way of further
+                    // clearing the buffer.  fail out and let the
+                    // caller know that we could not honor the
+                    // request
+                    SuccessOrExit(err);
+                    continue;
+                }
+                // we cannot copy event outright. We remember the
+                // current required space in mpAppData, we note the
+                // space requirements for the event in the current
+                // buffer and make that space in the next buffer.
+                eventBuffer->mAppData = reinterpret_cast<void *>(requiredSpace);
+                eventBuffer           = eventBuffer->mpNext;
+
+                // Sanity check: Die here on null event buffer.  If
+                // eventBuffer->mpNext were null, then the `EvictBuffer`
+                // in line 130 would have succeeded -- the event was
+                // already in the final buffer.
+                VerifyOrDie(eventBuffer != nullptr);
```

Comment:
Should this be an ASSERT if this is an internal sanity check?

I am thinking: either assert or return error, but not die.

Reply:
Done, thanks